### DISCUSSION TITLE
Milvus parent child data management

### DISCUSSION THREAD

**Discussion by molo6379** [milvus-io/milvus#32590] [Ideas & Feature requests]
Let's say I have a data that is in shape like this 
![image](https://github.com/milvus-io/milvus/assets/66596156/b7e028c2-e945-434c-811f-615cbc5a2e0d)

I want to make three different collection of [face, person, plate]

So if I want to push it into milvus DB the shape would be something like this -> (track_id, clip_emb(vector), best_img(path))
but then every track ID also has embedding vectors for themselves as well other than the representing clip_emb.

So in here the first (track_id, clip_emb(vector), best_img(path)) would be the parent and each track id has child which has their own embeddings and other components as well. 

Is it possible to mage data like this in Milvus?

**Comment by xiaofan-luan** [comment]
that sound's like an interesting use case.

What could the search looks like?

**Reply by xiaofan-luan** [reply]
So you have a tracking ID collection, with it's own embedding as reference for face, person and plate, this is the parent entity.
You also has face embeddings, person embeddings and plate embeddings?

When search happens, did you search from most similar faces and reverse lookup it's parent? or you search from trackings, and then from the list of faces search again for the most similar faces?

**Reply by molo6379** [reply]
What I was thinking were 2 use cases:

1. You do the search on child embeddings first and using the resulting top_ks of child embeddings, you can map its parent right in the same collection.

2. In case you want to search only on parent embeddings or child embeddings seperately for different senarios. 

   ex. clip_emb is used when you want to search image based on texts and embeddings for frames are used for image similarities. 
   So when user wants to seach image on text input, only clip_emb(parent) would be used for search but when the user wants to find similar images of certain image, only embeddings of each frames(child) would be used for that case

**Comment by xiaofan-luan** [comment]
1 -> currently you can use group by a field in your child embedding collection , see https://milvus.io/docs/single-vector-search.md#Grouping-search

2 ->  If we understand correctly, you are running embeddings for videos. Parent embedding is for the video itself, while daughter embeddings are for sub frames for the video. So sometimes you want to search for the entire video, and other times on single frame, is that correct?

I thought it's very similar to rag use cases. Sometimes you search by title and abstract and you have a other collection with chunked text and you search for the details

**Reply by molo6379** [reply]
1. Thank you I will try testing that!

2. Yes you are correct, and yes I think it is like rag use case. So sometimes I want to do search on parent and sometimes on children. Can it be done by group_search as well?

**Reply by molo6379** [reply]
+ 
```python
# Connect to Milvus
client = MilvusClient(uri='http://localhost:19530') # Milvus server address

# Load data into collection
client.load_collection("group_search") # Collection name

# Group search results
res = client.search(
    collection_name="group_search", # Collection name
    data=[[0.14529211512077012, 0.9147257273453546, 0.7965055218724449, 0.7009258593102812, 0.5605206522382088]], # Query vector
    search_params={
    "metric_type": "L2",
    "params": {"nprobe": 10},
    }, # Search parameters
    limit=10, # Max. number of search results to return
    group_by_field="doc_id", # Group results by document ID
    output_fields=["doc_id", "passage_id"]
)

# Retrieve the values in the `doc_id` column
doc_ids = [result['entity']['doc_id'] for result in res[0]]

print(doc_ids)
```
On this code, can I get a example collection of group_search? I have read the document how collection is formatted but I cant really visualize on my head,,, :(

**Reply by xiaofan-luan** [reply]
doc_ID field is actaully the parent field.

and here what we show is how to search child entities and get top related partent field.

parent has to be stored in another collection

**Reply by molo6379** [reply]
I see so it has two collections like [parent, children] and I can vector sim search on children embeddings and map it to the parent right? 

so lets say there's a data that looks like this: 
{parent_1:[child_1, child_2, child_3, child_4, child_5], parent_2:[child2_1, child2_2, child2_3, child2_4, child2_5]}
parent_1 and parent_2 have 5 children docs embeddings each then I need to make 2 different collections of parent and children right?
so something like this: 
Collection1(parent_1, parent_2) and Collection2(child_1, child_2, child_3, child_4, child_5, child2_1, child2_2, child2_3, child2_4, child2_5]) 

and if I want to grab parent doc according to vector sim search result of children collection, I should run search based on child doc embeddings and with the output result, I need to map it to the corresponding doc in parent collection?

If I am right on the procedure, how do I map it? Do I need to set the ID of children vectors with parent_id?

**Reply by xiaofan-luan** [reply]
search daughter , groupby parentID. and retrieve data from parentID

**Reply by molo6379** [reply]
So does each daughter entity has doc_id as well? such as:
Entity_1(emb_vector, passage_id_1, doc_id_1)
Entity_2(emb_vector, passage_id_2, doc_id_1)
Entity_3(emb_vector, passage_id_3, doc_id_1)

thanks for the quick answer btw :)

**Reply by xiaofan-luan** [reply]
that's the other id. 
the daugter entity can have both its own id and a parentID

**Comment by xiaofan-luan** [comment]
I do want to learn more this cross reference data model. It's kind like hard for a distributed database like milvus but it's also gonna to be very helpful for the developers
